internal/pkg/aws/sessions/sessions.go

```diff
@@ -29,6 +30,8 @@ const (
 	clientTimeout                   = 30 * time.Second
 )
 
+var ErrMissingRegion = errors.New("missing region configuration")
+
 // Provider provides methods to create sessions.
 // Once a session is created, it's cached locally so that the same session is not re-created.
 type Provider struct {
```

Comment:
I haven't read this completely ! so feel free to ignore the comment

I wonder if it would simplify anything if we do something like this:

```go
type Provider struct {
  ...
  logger func(format string, args ...interface{})
}


func NewProvider() *Provider {
  ...
  logger: log.Infof
}
```
and log the missing info stuff here. This way in the unit test we can have a logger that does actually nothing.

Reply:
I feel like the logging should still be happening from `cli` package. `NewProvider()` is called all over the place, for example `secretsmanager` and `config` packages. Whether we want the `logger` to be a `log.Infof` or a no-op only depends on `cli` though.  The result of this is that `secretsmanager` and `config`, being sometimes a dependency of `cli`, would then need to pass `cli`'s wills to `sessions` that whether the provider should have a `log.Infof` or a no-op as logger.

Reply:
We only log just before we return the error from `cli` - which means that we only log in the unit test if we are checking against the error. 

Expectedly, the log never appears in the unit test because we mock the provider such that the error wouldn't occur.

Reply:
I still think it's slightly problematic to have the logging in the `cli` pkg, because the moment we work on a new command we have to remember to add this log statement.

We have a few alternatives, but what do you think of this:
```go
package sessions

type ErrMissingRegion string
func (e *ErrMissingRegion) Error() string { return string(e) } // implements the error interface.

func (e *ErrMissingRegion) RecommendAction() string {  // implements new actionRecommender interface.
    return `It looks like your AWS region configuration is missing.
    ...`
}
```

Then in our main function for the binary: https://github.com/aws/copilot-cli/blob/28e91b9dc55fd0feed8bdf665679471843aa40d1/cmd/copilot/main.go#L25-L28

we do:

```go
package main
type actionRecommender interface {
   RecommendAction() string
}


if err := cmd.Execute(); err != nil {
  var ar actionRecommender
  if errors.As(err, &ar) {
    log.Infoln(ar.RecommendAction())
  }
  log.Errorln(err.Error())
  os.Exit(1)
}
```

Reply:
I thinks this pattern would also scale well to other errors